core/src/main/java/org/jdbi/v3/core/result/internal/ResultSetResultIterator.java

```diff
@@ -16,35 +16,56 @@
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.util.NoSuchElementException;
+import java.util.function.Supplier;
 
+import org.jdbi.v3.core.internal.exceptions.Sneaky;
 import org.jdbi.v3.core.mapper.RowMapper;
 import org.jdbi.v3.core.result.ResultIterator;
 import org.jdbi.v3.core.result.ResultSetException;
 import org.jdbi.v3.core.statement.StatementContext;
 
-import static java.util.Objects.requireNonNull;
-
 class ResultSetResultIterator<T> implements ResultIterator<T> {
     private final ResultSet resultSet;
     private final RowMapper<T> rowMapper;
+
+    private final ResultSetSupplier resultSetSupplier;
     private final StatementContext context;
 
     private volatile boolean alreadyAdvanced = false;
     private volatile boolean hasNext = false;
     private volatile boolean closed = false;
 
-    ResultSetResultIterator(ResultSet resultSet,
-                            RowMapper<T> rowMapper,
-                            StatementContext context) throws SQLException {
-        this.resultSet = requireNonNull(resultSet);
-        this.rowMapper = rowMapper.specialize(resultSet, context);
+    ResultSetResultIterator(Supplier<ResultSet> resultSetSupplier,
+        RowMapper<T> rowMapper,
+        StatementContext context) throws SQLException {
+
         this.context = context;
+
+        if (resultSetSupplier instanceof ResultSetSupplier) {
```

Comment:
Do we need this to be a dynamic instanceof check, or would it be good enough to have a static dispatch instead?

```
ResultSetResultIterator(ResultSetSupplier rs, ...) { ... }
ResultSetResultIterator(Supplier<ResultSet>rs, ...) {
    this(ResultSetSupplier.closingContext(rs, ctx);
}

Reply:
The code calling this will always call `ResultSetResultIterator(Supplier<ResultSet> ...)` independently whether the object is a ResultSetSupplier object or not. The problem is that between creating the ResultSetSupplier and its use there are a number of public APIs that expose `Supplier<ResultSet>`. Most prominently `ResultBearing.of` which is part of the public API.